Prisma Client sample app

ericharmeling opened the pull request:
Related to https://github.com/cockroachdb/docs/pull/12679.

This simple application demonstrates how to connect to CockroachDB with Prisma Client.

Commit:
Prisma sample app

otan commented on index.js:
```diff
@@ -0,0 +1,105 @@
+            await prisma.$queryRaw`COMMIT;`
+            return
+        } catch (err) {
+            if (err.code !== "40001") {
```
can you add a comment about what this error code is?

ericharmeling replied:
Done.

otan commented on index.js:
```diff
@@ -0,0 +1,105 @@
+    })
+    .finally(async () => {
+        await prisma.$disconnect()
+    })
```
wonder if we should have some auto js linter.

ericharmeling replied:
Just properly linted with eslinter. Were you saying we should add something automated to the repo?

otan replied:
automation would be cooler, but passing it through eslinter makes it look good!

Commit:
otan feedback

- Linted JS for style
- Added comment on SQLSTATE error

ericharmeling reviewed (commented):
TFTR @otan 

I'm going to pass this on to Prisma folks if you're good with it now.

millsp commented on index.js:
```diff
@@ -0,0 +1,101 @@
+
+// retryTxn wraps SQL operations in an explicit transaction.
+// If the transaction fails, the function retries the operation, with exponential backoff.
+async function retryTxn (n, max, operation) {
```
We have a similar feature over [here](https://www.prisma.io/docs/concepts/components/prisma-client/transactions#interactive-transactions-in-preview) (without retry logic).

Commit:
Updated table name to follow Prisma naming conventions

kernfeld-cockroach commented on dbinit.sql:
```diff
@@ -0,0 +1,4 @@
+CREATE TABLE "Account" (
+    id UUID PRIMARY KEY,
+    balance INT8
+);
```
Newline?

ericharmeling replied:
Done.

kernfeld-cockroach commented on index.js:
```diff
@@ -0,0 +1,101 @@
+
+// retryTxn wraps SQL operations in an explicit transaction.
+// If the transaction fails, the function retries the operation, with exponential backoff.
+async function retryTxn (n, max, operation) {
```
This shouldn't be the first snippet of code that a user sees. Here are my suggestions:

1. Just leave it out and mention retrying on, say, our Production Checklist
2. Use a library
3. Make the retry function more terse

ericharmeling replied:
Let's go with suggestion 1.

kernfeld-cockroach commented on index.js:
```diff
@@ -0,0 +1,101 @@
+async function initTable (rowVals) {
+  await prisma.account.createMany({
+    data: rowVals,
+    skipDuplicates: true
```
Could we omit `skipDuplicates`?

ericharmeling replied:
Done.

kernfeld-cockroach commented on index.js:
```diff
@@ -0,0 +1,101 @@
+}
+
+// initTable inserts the input row values into the account table.
+async function initTable (rowVals) {
```
I suggest using arrow functions to make us look "cool" and "modern."

ericharmeling replied:
The function wrappers are actually not needed if we aren't wrapping the operations in the retry wrapper.

kernfeld-cockroach commented on index.js:
```diff
@@ -0,0 +1,101 @@
+// deleteRows deletes all rows in the account table.
+async function deleteRows () {
+  const deleteAllRows = await prisma.account.deleteMany()
+  return console.log('Account rows deleted.', deleteAllRows)
```
Is `return console.log` intentional? I don't understand what that does.

ericharmeling replied:
Fixed.

kernfeld-cockroach commented on index.js:
```diff
@@ -0,0 +1,101 @@
+}
+
+// genVals generates random UUID values and integer values to be inserted into the account table.
+async function genVals (n) {
```
I'd prefer to make the business logic lighter-weight to make it easy for the user to focus on the code.

ericharmeling replied:
Done.

kernfeld-cockroach commented on index.js:
```diff
@@ -0,0 +1,101 @@
+  const accountValues = Array(n)
+  let i = 0
+  while (i < accountValues.length) {
+    const idVal = await uuidv4()
```
You shouldn't need to await UUID or random number generation. You can also inline `idVal` and `balVal` to shorten the code`.

ericharmeling replied:
Done.

kernfeld-cockroach commented on index.js:
```diff
@@ -0,0 +1,101 @@
+    const idVal = await uuidv4()
+    const balVal = await Math.floor(Math.random() * 1000)
+    accountValues[i] = { id: idVal, balance: balVal }
+    i++
```
I'd write this like: `const accountValues = Array(1000).fill().map(_ => ...)`

ericharmeling replied:
Done.

kernfeld-cockroach commented on index.js:
```diff
@@ -0,0 +1,101 @@
+}
+
+// genVals generates random UUID values and integer values to be inserted into the account table.
+async function genVals (n) {
```
I think this function doesn't need to be `async`.

ericharmeling replied:
Done.

kernfeld-cockroach commented on index.js:
```diff
@@ -0,0 +1,101 @@
+
+  await retryTxn(0, 15, initTable(firstAccount))
+
+  await new Promise(resolve => setTimeout(resolve, 2000)) // wait 2 seconds before selecting and printing rows
```
Can we remove this wait?

janpio replied:
(Interested in the response here as well, wonder why this is necessary)

ericharmeling replied:
I think the explicit transactions were causing concurrency issues with the `console` calls. I removed `retryTxn` and everything runs in the expected order. Removed the waits.

kernfeld-cockroach commented on package.json:
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "example-app-node-prisma",
+  "version": "1.0.0",
+  "description": "",
```
Maybe a simple description

ericharmeling replied:
Done.

kernfeld-cockroach commented on package.json:
```diff
@@ -0,0 +1,19 @@
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
```
You can remove the whole `scripts` block and other empty ones like `keywords`.

ericharmeling replied:
Done.

kernfeld-cockroach commented on package.json:
```diff
@@ -0,0 +1,19 @@
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
```
I think we do Apache 2 for everything

ericharmeling replied:
Done.

kernfeld-cockroach commented on package.json:
```diff
@@ -0,0 +1,19 @@
+  "name": "example-app-node-prisma",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
```
I think `main` is for when you're importing the code as a module, but this is instead to be used as a binary so maybe it makes more sense to use [`bin`](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#bin)?

ericharmeling replied:
Done.

kernfeld-cockroach reviewed (changes requested):
This is a great skeleton. I'd like to hammer this down to the simplest possible version.

janpio commented on index.js:
```diff
@@ -0,0 +1,101 @@
+const prisma = new PrismaClient()
+
+// retryTxn wraps SQL operations in an explicit transaction.
+// If the transaction fails, the function retries the operation, with exponential backoff.
```
A Prisma native way to implement this could be a middleware for example. Surprisingly that almost never comes up though, so does not exist in a nice version yet. 

What would the error states be that would be solved via a retry?

ericharmeling replied:
@janpio 
In production, [we recommend retrying queries that fail with SQLSTATE errors `40001`](https://www.cockroachlabs.com/docs/v21.2/transactions#client-side-intervention) (serialization error, usually due to a transaction failing because it conflicted with another concurrent or recent transaction accessing the same data). As @kernfeld-cockroach suggests, it might be best to just leave custom retry logic out of this sample app, and add a note in the tutorial about our [production recommendations](https://www.cockroachlabs.com/docs/stable/recommended-production-settings.html#transaction-retries).

janpio replied:
Thanks, that is interesting context. We could probably indeed do a suggested middleware for that if that is something Cockroach users would expect Prisma to handle in a nice way. (Although there might actually be a challenge about getting the right error code. Is there a way to _force_ such an error?)

kernfeld-cockroach replied:
Right, the retry is appropriate if and only if the error code is exactly 40001. 

It should be possible to deterministically generate a real error of this type by beginning one transaction, committing a conflicting transaction, and attempting to commit the first transaction.

In my opinion, the default user experience should be for the client to automatically retry a few times before giving up. Middleware might be a good solution, since power users might want to customize retry timing, which they could do by replacing the middleware. Whatever solution we decide to use for this with CockroachDB should probably also be used for Prisma's Postgres experience, since the 40001 error in Postgres has the [exact same semantics](https://www.postgresql.org/docs/current/transaction-iso.html).

janpio replied:
Prisma currently has the stance (by default) to not retry any queries, ever and let the user implement their own control flow (e.g. using middlewares or in their application directly). That way they have full control. 

On the other hand, this topic honestly almost never came up yet - especially not in context of `40001`. Surprising if you think about it...

janpio replied:
Now tracked at https://github.com/prisma/prisma/issues/11183 on our side.

2color commented:
Have you considered adding a relation? A bit part of Prisma Client's value is how easy it is to work with relations.

Commit:
App refactor

- Added new "Customer" table to database schema
- Added FK column to "Account" table, with ON DELETE CASCADE option
- Removed transaction retry wrapper
- Updated business logic to be more terse
- Updated node package file to be more complete, accurate
- Updated Prisma schema file with new database schema

ericharmeling reviewed (commented):
TFTR! I've refactored this sample app following the feedback. PTAL.

ericharmeling commented:
@2color 
> Have you considered adding a relation? A bit part of Prisma Client's value is how easy it is to work with relations.

Good idea! I added a very simple relation this sample app.

kernfeld-cockroach commented on dbinit.sql:
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE "Customer" (
+    id UUID PRIMARY KEY
```
I was thinking it would be more common to generate IDs using `DEFAULT gen_random_uuid()`. But then that would be more annoying to insert. @otan what do you think?

kernfeld-cockroach commented on index.js:
```diff
@@ -0,0 +1,50 @@
+
+main = async () => {
+  const customerIds = Array(10).fill().map(() => ({ id: uuidv4() }))
+  const accountValues = Array(10).fill().map((_, index) => ({ id: uuidv4(), customer_id: customerIds[index].id, balance: Math.floor(Math.random() * 1000) }))
```
This line of code is a spot lengthy for my tastes.

kernfeld-cockroach commented on index.js:
```diff
@@ -0,0 +1,50 @@
+
+const prisma = new PrismaClient()
+
+main = async () => {
```
`const main`?

kernfeld-cockroach commented on package.json:
```diff
@@ -0,0 +1,14 @@
+  "bin": "index.js",
+  "license": "Apache-2.0",
+  "devDependencies": {
+    "@prisma/client": "^3.7.0",
```
I think this will need to be a `dependencies` instead of a `devDependencies`.

kernfeld-cockroach reviewed (approved):
Nice!

Commit:
formatting/style updates
